# Exercise 3

## Part (a)

The $k$-fold cross-validation technique is implemented by dividing the data into $k$ groups called folds and then subsequently training whatever model or fit on $k-1$ fold while evaluating it on the held-out fold. We repeat this for all $k$ folds and return the cross-validation estimate as the average of all $k$ test errors.

## Part (b)

### Part (i) Validation set approach

The advatnages of $k$-fold cross-validation over the validation set approach are very apparent. First with $k$ folds we get a better estimate of the actual test error since we are averaging models trained on most of the data set but varying the held-out fold each time. The validation set approach is only trained once while leaving out the validation set entirely leading to a model with less training data and a higher error rate. 

### Part (ii) LOOCV

LOOCV has the opposite problem. First, since we are only leaving out one data point, we are training $n$ models which can be computationally epxensive for large data sets. However, since LOOCV is averaging so many models, it tends to give an unbiased estimate compared to $k$-fold CV. However, on the flip side, LOOCV has a much higher variance since each of the models that are being trained are almost identical to all the other model - with only 1 data point extra and missing between each model. Therefore there is a bias-variance tradeoff when it comes to $k$-fold CV (specifically a low value of $k$ such as 5 or 10) and LOOCV.

# Exercise 5

In [5]:
set.seed(314)
library(ISLR)

## Part (a)

In [6]:
log = glm(default~income+balance, data=Default, family="binomial")
summary(log)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = Default)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.154e+01  4.348e-01 -26.545  < 2e-16 ***
income       2.081e-05  4.985e-06   4.174 2.99e-05 ***
balance      5.647e-03  2.274e-04  24.836  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2920.6  on 9999  degrees of freedom
Residual deviance: 1579.0  on 9997  degrees of freedom
AIC: 1585

Number of Fisher Scoring iterations: 8


## Part (b)

### Part (i)

In [11]:
train = sample(nrow(Default), nrow(Default)/2)
tset = Default[train,]
vset = Default[-train,]

We are using a 50/50 split here!

### Part (ii)

In [16]:
train_log = glm(default~income+balance, data=tset, family="binomial")
summary(train_log)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = tset)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.148e+01  6.051e-01 -18.972  < 2e-16 ***
income       2.320e-05  6.876e-06   3.374 0.000742 ***
balance      5.598e-03  3.146e-04  17.794  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1497.19  on 4999  degrees of freedom
Residual deviance:  821.34  on 4997  degrees of freedom
AIC: 827.34

Number of Fisher Scoring iterations: 8


### Part (iii)

In [22]:
v_prob = predict(train_log, newdata=vset, type="response")
v_pred = ifelse(v_prob > 0.5, "Yes", "No")

### Part (iv)

In [23]:
mean(v_pred == vset$default)

[1] 0.9756

Validation Accuracy is 97.6% which is pretty good!

## Part (c)

In [25]:
train = sample(nrow(Default), 3*nrow(Default)/4)
tset = Default[train,]
vset = Default[-train,]
train_log = glm(default~income+balance, data=tset, family="binomial")
summary(train_log)
v_prob = predict(train_log, newdata=vset, type="response")
v_pred = ifelse(v_prob > 0.5, "Yes", "No")
mean(v_pred == vset$default)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = tset)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.163e+01  5.065e-01 -22.964  < 2e-16 ***
income       2.653e-05  5.809e-06   4.566 4.96e-06 ***
balance      5.597e-03  2.609e-04  21.453  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2192.2  on 7499  degrees of freedom
Residual deviance: 1199.3  on 7497  degrees of freedom
AIC: 1205.3

Number of Fisher Scoring iterations: 8


[1] 0.976

Validation Accuracy of 97.6% for a 75/25 train-valid split.

In [26]:
train = sample(nrow(Default), 3*nrow(Default)/5)
tset = Default[train,]
vset = Default[-train,]
train_log = glm(default~income+balance, data=tset, family="binomial")
summary(train_log)
v_prob = predict(train_log, newdata=vset, type="response")
v_pred = ifelse(v_prob > 0.5, "Yes", "No")
mean(v_pred == vset$default)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = tset)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.151e+01  5.496e-01 -20.948  < 2e-16 ***
income       2.365e-05  6.245e-06   3.788 0.000152 ***
balance      5.577e-03  2.862e-04  19.488  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1807.29  on 5999  degrees of freedom
Residual deviance:  983.18  on 5997  degrees of freedom
AIC: 989.18

Number of Fisher Scoring iterations: 8


[1] 0.9745

Validation Accuracy of 97.5% for a 60/40 train-valid split.

In [29]:
train = sample(nrow(Default), 4*nrow(Default)/5)
tset = Default[train,]
vset = Default[-train,]
train_log = glm(default~income+balance, data=tset, family="binomial")
summary(train_log)
v_prob = predict(train_log, newdata=vset, type="response")
v_pred = ifelse(v_prob > 0.5, "Yes", "No")
mean(v_pred == vset$default)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = tset)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.148e+01  4.789e-01 -23.977  < 2e-16 ***
income       2.308e-05  5.518e-06   4.183 2.88e-05 ***
balance      5.575e-03  2.479e-04  22.490  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2387.5  on 7999  degrees of freedom
Residual deviance: 1288.9  on 7997  degrees of freedom
AIC: 1294.9

Number of Fisher Scoring iterations: 8


[1] 0.9765

Validation Accuracy of 97.7% for a 80/20 train-valid split.

The accuracies for each split were actually pretty similar even when the splits were drastically different like the 50/50 and the 80/20 split. It may be simply due to random chance, but the percentages were not signficantly different.

## Part (d)

In [33]:
train = sample(nrow(Default), 4*nrow(Default)/5)
tset = Default[train,]
vset = Default[-train,]
train_log = glm(default~income+balance+student, data=tset, family="binomial")
summary(train_log)
v_prob = predict(train_log, newdata=vset, type="response")
v_pred = ifelse(v_prob > 0.5, "Yes", "No")
mean(v_pred == vset$default)


Call:
glm(formula = default ~ income + balance + student, family = "binomial", 
    data = tset)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.089e+01  5.476e-01 -19.889  < 2e-16 ***
income       6.110e-06  9.073e-06   0.673  0.50069    
balance      5.704e-03  2.584e-04  22.077  < 2e-16 ***
studentYes  -7.095e-01  2.646e-01  -2.682  0.00732 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2327.1  on 7999  degrees of freedom
Residual deviance: 1259.6  on 7996  degrees of freedom
AIC: 1267.6

Number of Fisher Scoring iterations: 8


[1] 0.9715

Using an 80/20 train-valid split, we get a validation accuracy (which in turn is our estimation for the test accuracy) of 97.2% which was actually lower than our accuracies above - albeit by a very small amount.

# Exercise 6

In [34]:
set.seed(218)

## Part (a)

In [35]:
log = glm(default~income+balance, data=Default, family="binomial")
summary(log)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = Default)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.154e+01  4.348e-01 -26.545  < 2e-16 ***
income       2.081e-05  4.985e-06   4.174 2.99e-05 ***
balance      5.647e-03  2.274e-04  24.836  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2920.6  on 9999  degrees of freedom
Residual deviance: 1579.0  on 9997  degrees of freedom
AIC: 1585

Number of Fisher Scoring iterations: 8


The estimated standard error for income is $4.99 \times 10^{-6}$ and $2.27\times10^{-4}$ for balance.

## Part (b)

In [37]:
boot.fn=function (data ,index)
    return(coef(glm(default~income+balance,data=data, family="binomial", subset=index)))

## Part (c)

In [40]:
library(boot)
boot(Default, boot.fn ,1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Default, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
         original        bias     std. error
t1* -1.154047e+01 -5.307360e-02 4.464786e-01
t2*  2.080898e-05  1.081510e-07 4.842164e-06
t3*  5.647103e-03  2.833599e-05 2.349243e-04

## Part (d)

The standard errors for income and balanace using the bootstrap method were $4.84 \times10^{-6}$ and $2.35\times10^{-4}$ both of which are close to the estimated values using the normal logistic regression, but slightly different. These values more accurately reflect the stnadrad errors in our data and don't make any unnecessary assumptions.